File: Example/BLECentralManager_Mac/InfoCharacteristic.m
//
//  InfoCharacteristic.m
//  BLECentralManager
//
//  Created by Balázs Kilvády on 6/21/16.
//  Copyright © 2016 Balázs Kilvády. All rights reserved.
//

#import "InfoCharacteristic.h"


@implementation InfoCharacteristic

- (instancetype)initWithName:(nonnull NSString *)name
{
    self = [super init];
    if (self) {
        _name = name;
    }
    return self;
}

- (void)device:(BLECDevice *)device
didFindCharacteristic:(CBCharacteristic *)characteristic
{
    DLog(@"device info characteristic <%@> found!", characteristic.UUID);
    [device.peripheral readValueForCharacteristic:characteristic];
}

- (void)device:(BLECDevice *)device
didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic
         error:(NSError *)error
{
    if (error) {
        DLog(@"characteristic value read with error: %@", error);
        return;
    }

    NSData *data = characteristic.value;
    NSString *value = [[NSString alloc] initWithBytes:[data bytes]
                                               length:[data length]
                                             encoding:NSUTF8StringEncoding];
    [_delegate infoCharacteristicName:_name value:value];
}

- (BOOL)device:(BLECDevice *)device releaseReadonlyCharacteristic:(CBCharacteristic *)characteristic
{
    return YES;
}

@end

File: Example/BLECentralManager_iOS_Swift/InfoCharacteristic.swift
//
//  InfoCharacteristic.swift
//  BLECentralManager_Mac_Swift
//
//  Created by Balázs Kilvády on 7/8/16.
//  Copyright © 2016 kil-dev. All rights reserved.
//

import BLECentralManager
import CoreBluetooth


class InfoCharacteristic: BLECCharacteristicDelegate {

    var delegate: InfoCharacteristicDelegate?
    private let _name: String

    init(name: String) {
        _name = name
    }

    func device(device: BLECDevice,
                didFindCharacteristic characteristic: CBCharacteristic) {
        DLog("device info characteristic <\(characteristic.UUID)> found!" )
        device.peripheral?.readValueForCharacteristic(characteristic)
    }


    func device(device: BLECDevice,
                didUpdateValueForCharacteristic characteristic: CBCharacteristic,
                error: NSError?) {
        guard error == nil else {
            DLog("characteristic value read with error: \(error)")
            return
        }
        guard let data = characteristic.value else {
            DLog("No data when it is updated?!")
            return
        }

        guard let value = String(data: data, encoding: NSUTF8StringEncoding) else {
            DLog("Cannot decode a string value?!")
            return
        }
        delegate?.infoCharacteristicName(_name, value: value)
    }

    func device(device: BLECDevice, releaseReadonlyCharacteristic characteristic: CBCharacteristic) -> Bool {
        return true
    }
}
